File: vignettes/02_tools/csharp/02_tools.cs
// Adding Tools to Agents (C#)
//
// This sample demonstrates how to define function tools and attach them
// to an agent so the model can call them automatically.
// It mirrors the Julia vignette 02_tools.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using System.ComponentModel;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// Define tool functions with [Description] attributes.
[Description("Get the current weather for a location.")]
static string GetWeather(
    [Description("The location to get the weather for.")] string location)
{
    string[] conditions = ["sunny", "cloudy", "rainy", "stormy"];
    var rng = new Random();
    return $"The weather in {location} is {conditions[rng.Next(conditions.Length)]} with a high of {rng.Next(10, 31)}°C.";
}

[Description("Get the approximate population of a country in millions.")]
static string GetPopulation(
    [Description("The country to look up.")] string country)
{
    var populations = new Dictionary<string, int>
    {
        ["France"] = 68,
        ["Germany"] = 84,
        ["Japan"] = 125,
        ["Brazil"] = 214,
        ["Australia"] = 26,
    };
    return populations.TryGetValue(country, out var pop)
        ? $"{country} has approximately {pop} million people."
        : $"Population data not available for {country}.";
}

[Description("Evaluate a mathematical expression and return the result.")]
static string Calculate(
    [Description("A mathematical expression to evaluate.")] string expression)
{
    // Simple evaluation via DataTable for demo purposes.
    var result = new System.Data.DataTable().Compute(expression, null);
    return result?.ToString() ?? "Error";
}

// Create the agent with tools.
AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        instructions: "You are a helpful assistant. Use the available tools to answer questions accurately. Be concise.",
        tools:
        [
            AIFunctionFactory.Create(GetWeather),
            AIFunctionFactory.Create(GetPopulation),
            AIFunctionFactory.Create(Calculate),
        ]);

// The agent will automatically call the GetWeather tool.
Console.WriteLine(await agent.RunAsync("What's the weather like in Tokyo?"));
Console.WriteLine();

// Ask about population — triggers the GetPopulation tool.
Console.WriteLine(await agent.RunAsync("What is the population of Brazil?"));
Console.WriteLine();

// A question that may require multiple tool calls.
Console.WriteLine(await agent.RunAsync("What is the combined population of France and Germany?"));

File: vignettes/02_tools/python/02_tools.py
"""
Adding Tools to Agents (Python)

This sample demonstrates how to define function tools with the @tool decorator
and attach them to an agent so the model can call them automatically.
It mirrors the Julia vignette 02_tools.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
from random import choice, randint
from typing import Annotated

from agent_framework import tool
from agent_framework.ollama import OllamaChatClient
from pydantic import Field


# Define tools using the @tool decorator.
# NOTE: approval_mode="never_require" is for sample brevity.
# Use "always_require" in production for user confirmation before tool execution.
@tool(approval_mode="never_require")
def get_weather(
    location: Annotated[str, Field(description="The location to get the weather for.")],
) -> str:
    """Get the current weather for a location."""
    conditions = ["sunny", "cloudy", "rainy", "stormy"]
    return f"The weather in {location} is {choice(conditions)} with a high of {randint(10, 30)}°C."


@tool(approval_mode="never_require")
def get_population(
    country: Annotated[str, Field(description="The country to look up.")],
) -> str:
    """Get the approximate population of a country in millions."""
    populations = {
        "France": 68,
        "Germany": 84,
        "Japan": 125,
        "Brazil": 214,
        "Australia": 26,
    }
    pop = populations.get(country)
    if pop is not None:
        return f"{country} has approximately {pop} million people."
    return f"Population data not available for {country}."


@tool(approval_mode="never_require")
def calculate(
    expression: Annotated[str, Field(description="A mathematical expression to evaluate.")],
) -> str:
    """Evaluate a mathematical expression and return the result."""
    # WARNING: eval is used here for demo purposes only.
    result = eval(expression)  # noqa: S307
    return str(result)


async def main() -> None:
    # Create a chat client and agent with tools.
    client = OllamaChatClient(
        host="http://localhost:11434",
        model_id="qwen3:8b",
    )

    agent = client.as_agent(
        name="ToolAgent",
        instructions="You are a helpful assistant. Use the available tools to answer questions accurately. Be concise.",
        tools=[get_weather, get_population, calculate],
    )

    # The agent will automatically call the get_weather tool.
    result = await agent.run("What's the weather like in Tokyo?")
    print(f"Agent: {result}\n")

    # Ask about population — triggers the get_population tool.
    result = await agent.run("What is the population of Brazil?")
    print(f"Agent: {result}\n")

    # A question that may require multiple tool calls.
    result = await agent.run("What is the combined population of France and Germany?")
    print(f"Agent: {result}")


if __name__ == "__main__":
    asyncio.run(main())
